JuiceShop selenium auth example (#297)
zaproxy/community-scripts: af-plans/juiceshop-selenium-auth/README.md

# Juice Shop Selenium Authentication

> This directory contains [Automation Framework](https://www.zaproxy.org/docs/automate/automation-framework/) yaml plans and a set of scripts to demonstrate how to set up ZAP to automate authentication using Selenium to launch browsers.

Reviewer: `YAML` and in other places.

Note that ZAP _can_ authenticate to Juice Shop _without_ using Selenium, and normally this is what we would recommend.

> However in this case we are using it to show how to use Selenium to authenticate to your app if there is no alternative.

Reviewer 2: ```suggestion
However, in this case we are using it to show how to use Selenium to authenticate to your app if there is no alternative.
```

## Requirements

To run this plan you need to:

* Install Java 11 - unfortunately we currently have to use Oracle Nashorn for the Selenium scripting which is not available in later versions of Java
* Install ZAP

> * Run [Juice Shop](https://github.com/juice-shop/juice-shop) so that iit is accessible via http://localhost:3000/

Reviewer 2: ```suggestion
* Run [Juice Shop](https://github.com/juice-shop/juice-shop) so that it is accessible via http://localhost:3000/
```

* Add a user to Juice Shop with the credentials:
  * username: test@test.com
  * password: test123
* Edit all of the yaml files and shell scripts to replace `/full/path/` to be the local full path of your scripts
* Edit all of the shell scripts to correct the path to your ZAP installation
* Ensure that the user credentials are available via the environmental variables (these _should_ be set up correctly via the supplied scripts):
  * JS_USER: test@test.com
  * JS_PWD: test123

## Test Run

> With the above requirements in place run the script `./js-test.js` (Linux or MacOS) or `js-test.bat` (Windows).

Reviewer: `macOS` and other occurrences.

This uses the `juiceshop-test-req.yaml` file which sets up the environment and just makes one authenticated request.

> A set of [Job Tests](https://www.zaproxy.org/docs/desktop/addons/automation-framework/tests/) check that the authenticate worked as expected.

Reviewer: `authentication`

> You should see messages on the command line like:

Reviewer 2: ```suggestion
You should see messages in your console similar to:
```

```
Job requestor test of type stats passed: At least 1 successful login [1 >= 1]
Job requestor test of type stats passed: No login failures [0 <= 0]
Job requestor finished
Automation plan succeeded!
```

These indicate that the plan ran successfully.

> Note that Firefox reports lots of messages to the command line - these are difficult to suppress :/

Reviewer 2: ```suggestion
Note that Firefox reports lots of messages to the console - these are difficult to suppress :/
```

To show that ZAP fails the plan if the authentication fails, either do not register the user or change the script to use different credentials.

## Full Run

With the above requirements in place run the script `./js-auth.js` (Linux or MacOS) or `js-auth.bat` (Windows).

> This uses the `juiceshop-auth.yaml` file which sets up the environment and runs both the standard and ajax spider,

Reviewer 2: ```suggestion
This uses the `juiceshop-auth.yaml` file which sets up the environment and runs both the standard and AJAX Spider,
```

both of which authenticate using the `test@test.com` user.

> A set of [Job Tests](https://www.zaproxy.org/docs/desktop/addons/automation-framework/tests/) check that the authenticate worked as expected for both spiders.

Reviewer: `authentication`

> You should see messages on the command line like:

Reviewer 2: ```suggestion
You should see messages in the console like:
```

Author: Couldnt that be read as the ZAP script console? In this case these messages will be writen to stdout..

Author: I'm all for getting it right 😁

Author: @thc202 has suggested "standard output" as it depends how ZAP is launched - that works for me..

Reviewer 2: Okay, lets go that way

```
Job spiderAjax found 1,898 URLs
Job spiderAjax test of type stats passed: At least 1000 URLs found [1898 >= 1000]
Job spiderAjax test of type stats passed: At least 1 successful login [1 >= 1]
Job spiderAjax test of type stats passed: No login failures [0 <= 0]
Job spiderAjax test of type stats passed: At least 300 authentication tokens present [374 >= 300]
Job spiderAjax test of type stats failed: Less than 500 authentication tokens absent [1051 >= 500]
Job spiderAjax test of type stats passed: At least 300 authentication cookies present [930 >= 500]
Job spiderAjax test of type stats passed: Less than 500 authentication cookies absent [495 < 500]
Job spiderAjax finished
Job passiveScan-wait started
Job passiveScan-wait finished
Automation plan succeeded!
```

These indicate that the plan ran successfully.

> Note that Firefox reports lots of messages to the command line - these are difficult to suppress :/

Reviewer 2: ```suggestion
Note that Firefox reports lots of messages via the console - these are difficult to suppress :/
```
